### ISSUE TITLE
em++ - crash on c++17 static inline initialization

### ISSUE THREAD

**Issue by gl84** [emscripten-core/emscripten#14025]
em++ asserts on c++17 static inline initializations:
```
wasm-ld: /b/s/w/ir/cache/builder/emscripten-releases/llvm-project/lld/wasm/Symbols.cpp:305: uint32_t lld::wasm::GlobalSymbol::getGlobalIndex() const: Assertion `globalIndex != INVALID_INDEX' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
 #0 0x00007f2ce35eee93 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x1925e93)
 #1 0x00007f2ce35eccce llvm::sys::RunSignalHandlers() (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/../lib/libLLVM-13git.so+0x1923cce)
 #2 0x00007f2ce35ef35f SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f2ce5bd03c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #4 0x00007f2ce19b318b raise /build/glibc-eX1tMB/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f2ce1992859 abort /build/glibc-eX1tMB/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f2ce1992729 get_sysdep_segment_value /build/glibc-eX1tMB/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f2ce1992729 _nl_load_domain /build/glibc-eX1tMB/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007f2ce19a3f36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
 #9 0x0000000000728b9e (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/wasm-ld+0x728b9e)
#10 0x0000000000710473 lld::wasm::ObjFile::calcNewValue(llvm::wasm::WasmRelocation const&, unsigned long, lld::wasm::InputChunk const*) const (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/wasm-ld+0x710473)
#11 0x000000000070cda8 lld::wasm::InputChunk::writeTo(unsigned char*) const (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/wasm-ld+0x70cda8)
#12 0x000000000074bc3d lld::wasm::CustomSection::writeTo(unsigned char*) (/home/meiner/00_work/20_Programs/emscripten/emsdk_tot/emsdk/upstream/bin/wasm-ld+0x74bc3d)
#13 0x0000000000740abd std::_Function_handler<void (), void llvm::parallel::detail::parallel_for_each<__gnu_cxx::__normal_iterator<lld::wasm::OutputSection**, std::vector<lld::wasm::OutputSection*, std::allocator<lld::wasm::OutputSection*> > >, lld::wasm::(anonymous namespace)::Writer::writeSections()::$_0>(__gnu_cxx::__normal_iterator<lld::wasm::OutputSection**, std::vector<lld::wasm::OutputSection*, std::allocator<lld::wasm::OutputSection*> > >, __gnu_cxx::__normal_iterator<lld::wasm::OutputSection**, std::vector<lld::wasm::OutputSection*, std::allocator<lld::wasm::OutputSection*> > >, lld::wasm::(anonymous namespace)::Writer::writeSections()::$_0)::'lambda'()>::_M_invoke(std::_Any_data const&) Writer.cpp:0:0
#14 0x00007f2ce3574c85 std::_Function_handler<void (), llvm::parallel::detail::TaskGroup::spawn(std::function<void ()>)::$_0>::_M_invoke(std::_Any_data const&) Parallel.cpp:0:0
#15 0x00007f2ce35744c1 llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::work(llvm::ThreadPoolStrategy, unsigned int) Parallel.cpp:0:0
#16 0x00007f2ce58ac9ef execute_native_thread_routine WindowsManifestMerger.cpp:0:0
#17 0x00007f2ce5bc4609 start_thread /build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:478:7
#18 0x00007f2ce1a8f293 __clone /build/glibc-eX1tMB/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97:0
```

A simple repro case using c++17 static inline initialization is:
```c++
struct cpp17_static_member_var {
    static inline int* var = new int;
};
```
Note: To trigger the assertion both **SIDE_MODULE=2** and **-g** compilation flags must be used, e.g.
 `em++ -std=c++17 -s SIDE_MODULE=2 -g side.cpp`

**Comment by sbc100** [comment]
Fix is in https://reviews.llvm.org/D101399

**Comment by gl84** [comment]
Amazing how fast you found it!

Thanks a lot!

**Comment by sbc100** [comment]
Actually I can't reproduce this with just a function local static without c++17:

```
$ cat test.cc 
void* foo() {
    static int* var = new int;
    return var;
}
$ ./em++ -g test.cc -s SIDE_MODULE=2
wasm-ld: /usr/local/google/home/sbc/dev/wasm/llvm-project/lld/wasm/Symbols.cpp:305: uint32_t lld::wasm::GlobalSymbol::getGlobalIndex() const: Assertion `globalIndex != INVALID_INDEX' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
...
```

I'm still trying to understand why this is only happening with `-s SIDE_MODULE=2`

**Comment by sbc100** [comment]
Actually it reproduces with just:

```
int foo() {
    return 0;
}
```

Looks like the problem is with any module small enough not to use the stack pointer.. because `SIDE_MODULE=1` builds with `--gc-sections` that `__stack_pointer` global is GC'd in that case.

**Comment by gl84** [comment]
Even better. Sounds like this could also fix my old issue #10983.

Thanks!

**Comment by sbc100** [comment]
At yes is the same issue I think.  Closing as duplicate of #10983.
